### PR TITLE
Adding parameter to specify more master SANs

### LINKED ISSUE #2614: Provision to specify SANs to add in API server certificate
**Is this a request for help?**:
NO

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
FEATURE

---

**What version of acs-engine?**:
0.15.1

---

Provision to define other master SAN(s) in apimodel could be potentially useful for cluster administrators when migrating applications etc. to a new cluster.

For instance, when we need to upgrade a Kubernetes environment, we create a new cluster and migrate all applications over to new cluster, then point our DNS to the public IP (of ingress controller) of new cluster.
But we are not able to do this for API server domain because it's tied with the master FQDN specified in the apimodel during cluster creation (via the API server certificate). We can't use same master FQDN for 2 different cluster in same region because they need to be different because of limitations in design+Azure.

If we were able to specify master SAN(s) and add them to API server certificate, we could create multiple clusters with different master FQDN but **same** SAN and swap between them using an external domain name.

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**:
N/A

**What you expected to happen**:
Described above

**How to reproduce it** (as minimally and precisely as possible):
N/A

**Anything else we need to know**:
For reference below are the SANs currently created by acs-engine in the API server certificate (for masterFQDN=`temp`)
```
DNS:temp.australiaeast.cloudapp.azure.com, DNS:temp.australiasoutheast.cloudapp.azure.com, DNS:temp.brazilsouth.cloudapp.azure.com, DNS:temp.canadacentral.cloudapp.azure.com, DNS:temp.canadaeast.cloudapp.azure.com, DNS:temp.centralindia.cloudapp.azure.com, DNS:temp.centralus.cloudapp.azure.com, DNS:temp.centraluseuap.cloudapp.azure.com, DNS:temp.chinaeast.cloudapp.chinacloudapi.cn, DNS:temp.chinanorth.cloudapp.chinacloudapi.cn, DNS:temp.eastasia.cloudapp.azure.com, DNS:temp.eastus.cloudapp.azure.com, DNS:temp.eastus2.cloudapp.azure.com, DNS:temp.eastus2euap.cloudapp.azure.com, DNS:temp.japaneast.cloudapp.azure.com, DNS:temp.japanwest.cloudapp.azure.com, DNS:temp.koreacentral.cloudapp.azure.com, DNS:temp.koreasouth.cloudapp.azure.com, DNS:temp.northcentralus.cloudapp.azure.com, DNS:temp.northeurope.cloudapp.azure.com, DNS:temp.southcentralus.cloudapp.azure.com, DNS:temp.southeastasia.cloudapp.azure.com, DNS:temp.southindia.cloudapp.azure.com, DNS:temp.uksouth.cloudapp.azure.com, DNS:temp.ukwest.cloudapp.azure.com, DNS:temp.westcentralus.cloudapp.azure.com, DNS:temp.westeurope.cloudapp.azure.com, DNS:temp.westindia.cloudapp.azure.com, DNS:temp.westus.cloudapp.azure.com, DNS:temp.westus2.cloudapp.azure.com, DNS:temp.chinaeast.cloudapp.chinacloudapi.cn, DNS:temp.chinanorth.cloudapp.chinacloudapi.cn, DNS:temp.germanycentral.cloudapp.microsoftazure.de, DNS:temp.germanynortheast.cloudapp.microsoftazure.de, DNS:temp.usgovvirginia.cloudapp.usgovcloudapi.net, DNS:temp.usgoviowa.cloudapp.usgovcloudapi.net, DNS:temp.usgovarizona.cloudapp.usgovcloudapi.net, DNS:temp.usgovtexas.cloudapp.usgovcloudapi.net, DNS:temp.francecentral.cloudapp.azure.com, DNS:kubernetes, DNS:kubernetes.default, DNS:kubernetes.default.svc, DNS:kubernetes.default.svc.cluster.local, DNS:kubernetes.kube-system, DNS:kubernetes.kube-system.svc, DNS:kubernetes.kube-system.svc.cluster.local, IP Address:10.255.255.5, IP Address:10.255.255.15, IP Address:10.0.0.1
```

### PR DESCRIPTION
**Pull request by aku105**
**What this PR does / why we need it**: Adding a new parameter to define more SAN's using which API servers can be reached using kubectl.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2614

**Special notes for your reviewer**:

**If applicable**:
- [x] documentation
- [x] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -309,6 +309,7 @@ type DcosConfig struct {
 type MasterProfile struct {
 	Count                    int               `json:"count"`
 	DNSPrefix                string            `json:"dnsPrefix"`
+	MasterSANs               []string          `json:"masterSANs"`
```
Comment (jackfrancis) [inline pkg/api/types.go]: Let's change this property to `SubjectAltNames`, just to be a little clearer what it is (remove acronym if possible :))

File: docs/clusterdefinition.md [hunk 2]
```diff
@@ -419,6 +419,7 @@ We consider `kubeletConfig`, `controllerManagerConfig`, `apiServerConfig`, and `
 |---|---|---|
 |count|yes|Masters have count value of 1, 3, or 5 masters|
 |dnsPrefix|yes|The dns prefix for the master FQDN.  The master FQDN is used for SSH or commandline access. This must be a unique name. ([bring your own VNET examples](../examples/vnet))|
+|masterSANs|no|An array of domain names using which a user can reach API server. These domains are added as SAN to the generated API server certificate. **NOTE**: These domains **will not** be automatically provisioned.|
```
Comment (jackfrancis) [inline docs/clusterdefinition.md]: replace "SAN" with "Subject Alternative Name" for clarity

**Comment by aku105** [comment]
@jackfrancis Updated the variable names and docs to improve readability, as suggested.

**Comment by jackfrancis** [comment]
@CecileRobertMichon @slack @amanohar  would appreciate a high level review of the intent here and implementation, because the `MasterProfile` type struct is general purpose. I'm in favor of this addition.

**Comment by jackfrancis** [comment]
@aku105 Let's add a unit test that proves out the following usage patterns:

- nil `MasterProfile.SubjectAltNames` input produces certs w/ the dnsPrefix only
- populated `MasterProfile.SubjectAltNames` produces certs w/ the dnsPrefix string + the add'l SANs in the input

**Comment by aku105** [comment]
Hi @jackfrancis I've added an unit test.
Could you please confirm if this is what you were expecting in the unit test?

**Pull request merged**